src/envs/pybullet_blocks.py

```diff
@@ -0,0 +1,728 @@
+            current_val = state.get(self._robot, "fingers")
+            gripper_action = target_val - current_val
+            gripper_action = np.clip(gripper_action, self.action_space.low[3],
+                                     self.action_space.high[3])
+            return Action(
+                np.array([0., 0., 0., gripper_action], dtype=np.float32))
+
+        def _terminal(state: State, memory: Dict, objects: Sequence[Object],
+                      params: Array) -> bool:
+            del memory, objects, params  # unused
+            current_val = state.get(self._robot, "fingers")
+            dist = (target_val - current_val)**2
+            return dist < self._grasp_tol
+
+        return ParameterizedOption(name,
+                                   types=types,
+                                   params_space=params_space,
+                                   policy=_policy,
+                                   initiable=lambda _1, _2, _3, _4: True,
+                                   terminal=_terminal)
+
+    def _move_relative_to_table(self, name: str,
+                                rel_or_abs_target_pose: Pose3D,
+                                rel_or_abs: Tuple[str, str,
+                                                  str], types: Sequence[Type],
+                                params_space: Box) -> ParameterizedOption:
+        """Creates a ParameterizedOption for moving to a target pose relative
+        to the table.
+
+        Each of the three dimensions of the target pose can be specified
+        relatively or absolutely. The rel_or_abs argument indicates
+        whether each dimension is relative ("rel") or absolute ("abs").
+        """
+
+        def _get_current_and_target_pose(
+                state: State, objects: Sequence[Object],
+                params: Array) -> Tuple[Pose3D, Pose3D]:
+            robot, = objects
+            # De-normalize parameters to actual table coordinates.
+            x_norm, y_norm = params
+            x = self.x_lb + (self.x_ub - self.x_lb) * x_norm
+            y = self.y_lb + (self.y_ub - self.y_lb) * y_norm
+            current_pose = (state[robot][0], state[robot][1], state[robot][2])
+            table_pose = (x, y, self.table_height)
+            target_pose = self._convert_rel_abs_to_abs(table_pose,
+                                                       rel_or_abs_target_pose,
+                                                       rel_or_abs)
+            return current_pose, target_pose
+
+        return self._move_to_pose(name, types, params_space,
+                                  _get_current_and_target_pose)
+
+    def _get_move_toward_waypoint_action(self, gripper_pose: Pose3D,
+                                         target_pose: Pose3D) -> Action:
+        action = self._move_gain * np.subtract(target_pose, gripper_pose)
+        action_norm = np.linalg.norm(action)  # type: ignore
+        if action_norm > self._max_vel_norm:
```

Comment:
how often does this occur in practice, given how tiny our action space bounds are?

Reply:
the difference between target_pose and gripper_pose can be very large, so this occurs all the time